src/EditorFeatures/Core/Implementation/PasteTracking/PasteTrackingService.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PasteTracking
+{
+    [Export(typeof(IPasteTrackingService)), Shared]
+    [Export(typeof(PasteTrackingService))]
+    internal class PasteTrackingService : IPasteTrackingService
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly object _pastedTextSpanKey = new object();
+
+        [ImportingConstructor]
+        public PasteTrackingService(IThreadingContext threadingContext)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public bool TryGetPastedTextSpan(Document document, out TextSpan textSpan)
+        {
+            if (!TryGetTextBuffer(document, out var textBuffer))
+            {
+                textSpan = default;
+                return false;
+            }
+
+            // `PropertiesCollection` is thread-safe
+            return textBuffer.Properties.TryGetProperty(_pastedTextSpanKey, out textSpan);
+        }
+
+        internal void RegisterPastedTextSpan(ITextBuffer textBuffer, TextSpan textSpan)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            textBuffer.Changed += RemovePastedTextSpan;
+            textBuffer.Properties.AddProperty(_pastedTextSpanKey, textSpan);
+
+            return;
+
+            void RemovePastedTextSpan(object sender, TextContentChangedEventArgs e)
+            {
+                textBuffer.Changed -= RemovePastedTextSpan;
+                textBuffer.Properties.RemoveProperty(_pastedTextSpanKey);
+            }
```

Comment:
this all looks good to me.  but paging @jasonmalinowski in case he has any thoughts here about anything to be concerned about.